api/http/management.go

```diff
@@ -371,23 +390,53 @@ Loop:
 				return err
 			}
 			if mr.Header.Proto == ws.ProtoTypeShell {
-				if mr.Header.MsgType == shell.MessageTypeShellCommand {
+				switch mr.Header.MsgType {
+				case shell.MessageTypeShellCommand:
+					timeNowUTC := int64(
+						float64(time.Now().UTC().UnixNano()) * 0.000001)
+					keystrokeDelay := timeNowUTC - lastKeystrokeAt
+					if keystrokeDelay >= keyStrokeDelayRecordingThresholdMs {
+						controlMsg := app.Control{
+							Type:           app.DelayMessage,
+							Offset:         recordedBytes,
+							DelayMs:        uint16(keystrokeDelay),
+							TerminalHeight: 0,
+							TerminalWidth:  0,
+						}
+						n, _ := controlRecorderBuffered.Write(
+							controlMsg.GetBytes())
+						l.Debugf("saving control delay message: %+v/%d",
+							controlMsg, n)
+						controlBytes += n
+						if controlBytes >= app.MessageSizeLimit {
+							l.Infof("closing session session_id=%s"+
+								"on control data limit reached.",
+								session.ID)
+							//see MEN-4448
+							break Loop
+						}
+					}
+					lastKeystrokeAt = timeNowUTC
 					b, e := recorderBuffered.Write(mr.Body)
 					if e != nil {
 						l.Errorf("session logging: "+
 							"recorderBuffered.Write(len=%d)=%d,%+v",
 							len(mr.Body), b, e)
 					}
 					recordedBytes += len(mr.Body)
+					session.BytesRecordedMutex.Lock()
+					session.BytesRecorded = recordedBytes
+					session.BytesRecordedMutex.Unlock()
 					if recordedBytes >= app.MessageSizeLimit {
 						l.Infof("closing session on limit reached.")
 						//see https://tracker.mender.io/browse/MEN-4448
 						break Loop
 					}
-				} else if mr.Header.MsgType == shell.MessageTypeStopShell {
+				case shell.MessageTypeStopShell:
```

Comment:
If I end the session by closing the connection; does that mean the session never gets recorded?

Reply:
I added some flushes in case of error and in other places where this loop is broken. I any case what ever got flushed would be recorded. now I think it would be even harder to bypass the recording.

Reply:
The defer statement ^ should be enough. You can delete all other Flush calls.